lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+	}
+	txnData.ExpirationTimestampUnixNanoSecs = int64(uint64ExpirationTimestampUnixNanoSecs)
+
+	// LockedDAOCoinToTransferBaseUnits
+	txnData.LockedCoinsToTransferBaseUnits, err = VariableDecodeUint256(rr)
+	if err != nil {
+		return errors.Wrapf(err, "DAOCoinLockupTransferMetadata.FromBytes: Problem reading LockedDAOCoinToTransferBaseUnits")
+	}
+
+	return nil
+}
+
+func (txnData *CoinLockupTransferMetadata) New() DeSoTxnMetadata {
+	return &CoinLockupTransferMetadata{}
+}
+
+//
+// TYPES: CoinUnlockMetadata
+//
+
+type CoinUnlockMetadata struct {
+	ProfilePublicKey       *PublicKey
+	CoinsToUnlockBaseUnits *uint256.Int
+}
+
+func (txnData *CoinUnlockMetadata) GetTxnType() TxnType {
+	return TxnTypeCoinUnlock
+}
+
+func (txnData *CoinUnlockMetadata) ToBytes(preSignature bool) ([]byte, error) {
+	var data []byte
+	data = append(data, EncodeByteArray(txnData.ProfilePublicKey.ToBytes())...)
+	data = append(data, VariableEncodeUint256(txnData.CoinsToUnlockBaseUnits)...)
+	return data, nil
+}
+
+func (txnData *CoinUnlockMetadata) FromBytes(data []byte) error {
+	rr := bytes.NewReader(data)
+
+	// ProfilePublicKey
+	profilePublicKeyBytes, err := DecodeByteArray(rr)
+	if err != nil {
+		return errors.Wrapf(err, "CoinUnlockMetadata.FromBytes: Problem reading ProfilePublicKey")
+	}
+	txnData.ProfilePublicKey = NewPublicKey(profilePublicKeyBytes)
+
+	// CoinToUnlockBaseUnits
+	txnData.CoinsToUnlockBaseUnits, err = VariableDecodeUint256(rr)
+	if err != nil {
+		return errors.Wrapf(err, "CoinUnlockMetadata.FromBytes: Problem reading DAOCoinToUnlockBaseUnits")
+	}
+
+	return nil
+}
+
+func (txnData *CoinUnlockMetadata) New() DeSoTxnMetadata {
+	return &CoinUnlockMetadata{}
+}
+
+//
+// CoinLockup Transaction Logic
+//
+
+func (bav *UtxoView) _connectCoinLockup(
+	txn *MsgDeSoTxn, txHash *BlockHash,
+	blockHeight uint32, blockTimestamp int64,
+	verifySignatures bool) (_totalInput uint64,
+	_totalOutput uint64, _utxoOps []*UtxoOperation, _err error) {
+	var utxoOpsForTxn []*UtxoOperation
+
+	// Validate the starting block height.
+	if blockHeight < bav.Params.ForkHeights.ProofOfStake1StateSetupBlockHeight ||
+		blockHeight < bav.Params.ForkHeights.BalanceModelBlockHeight {
+		return 0, 0, nil,
+			errors.Wrapf(RuleErrorLockupTxnBeforeBlockHeight, "_connectCoinLockup")
+	}
+
+	// Validate the txn TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeCoinLockup {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectCoinLockup: called with bad TxnType %s", txn.TxnMeta.GetTxnType().String(),
+		)
+	}
+
+	// Try connecting the basic transfer without considering transaction metadata.
+	_, _, utxoOpsForBasicTransfer, err := bav._connectBasicTransfer(txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectCoinLockup")
+	}
+	utxoOpsForTxn = append(utxoOpsForTxn, utxoOpsForBasicTransfer...)
+
+	// Grab the txn metadata.
+	txMeta := txn.TxnMeta.(*CoinLockupMetadata)
```

Comment:
So something a little bit tricky is that we have global transfer restrictions for a coin that are supposed to disallow people from sending coins to each other before we're ready. I think you need to check that "non-lockup" transfer restriction in this function. It should be pretty easy though. You just have to call a function like IsValidTransfer or something that Isaac wrote a while back.